Fixes #37566 - Libvirt - UEFI & SecureBoot support

stejskalleos opened the pull request:
Required fog-libvirt PR: https://github.com/fog/fog-libvirt/pull/155
<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

stejskalleos committed:
Fixes #37566 - Libvirt - UEFI & SecureBoot support

Co-authored-by: Ewoud Kohl van Wijngaarden <ewoud@kohlvanwijngaarden.nl>

stejskalleos commented on app/models/compute_resources/foreman/model/libvirt.rb:
```diff
@@ -14,7 +14,35 @@ def self.available?
+
+    def self.os_loader_defaults
+      {
+        secure: "no",
```
While testing secure boot with Fedora 39, I discovered that when using `shimx64.efi`, the secure boot is always enabled, ignoring the `loader secure=no`.

[From the docs:](https://libvirt.org/formatdomain.html)
> Attribute secure can be used to tell the hypervisor that the firmware is capable of Secure Boot feature. It cannot be used to enable or disable the feature itself in the firmware.

So now I'm not sure how to handle this in UI. Should we keep it as it is?

ares replied:
What do you mean by secure boot being enabled? Does it enforce the signature verification? Or shim just boots grub regardless of the signature? How would libvirt know that shim was booted and enabled secure boot in uefi accordingly?

I'm just trying to understand, I guess giving such options to create the VM makes sense anyways. We could also reach out to libvirt devs to find out the intentended behavior.

stejskalleos replied:
If you use `shimx64.efi` as a bootloader and set the  `secure=no` in UI form, the VM will start with enabled secure boot.

It might be confusing for the users in the same way as it was for me.
I think the label should be `Secure Boot Capable` instead of `Secure Boot`.
Or add some help text explaining what is it exactly doing.

jloeser replied:
As the docs says, Secure Boot is not enabled by just booting `shimx64.efi`. This is basically configured in the firmware.

IIRC most distributions always boot via `shimx64.efi`. If SecureBoot is not enabled in firmware, shim simply loads the bootloader (GRUB2) and that's it, no signature check at all.

According to https://libvirt.org/kbase/secureboot.html the interesting option is:
```
  <firmware>
    <feature enabled='yes' name='secure-boot'/>
    <feature enabled='yes' name='enrolled-keys'/>
  </firmware>
```

We would need to check what the actual fog-libvirt plugin (I assume) is doing when talking to libvirt API and to make sure, that corresponding (XML) options are provided when Secure Boot is enabled in the UI or not.

stejskalleos replied:
Oh, so enabling secure_boot in UI should manage these firmware features ... I was thinking about them, where they are coming from; I thought they are set by libvirt itself, not by the fog. Hmm, will take a look at it.

jloeser replied:
I had a quick look and I think here is a good point to start: https://github.com/fog/fog-libvirt/blob/21d1bf827c03ae79e1f5a10118c2b41fb80e0be7/lib/fog/libvirt/models/compute/server.rb#L271

This looks like the place where the domain/server XML definition gets assembled which is then send to libvirt API.

jloeser replied:
Looks like there was already some work done by you guys https://github.com/fog/fog-libvirt/pull/134 (at least adding UEFI firmware).

stejskalleos replied:
yop, this is my follow up PR https://github.com/fog/fog-libvirt/pull/141

jloeser replied:
So just for clarification (I didn't test by myself):

Is SB enabled or disabled  in the firmware depending `secure` variable in combination with https://github.com/fog/fog-libvirt/pull/141 or is it always enabled even if `secure=no`?

stejskalleos replied:
It's always enabled. `secure=no` is just telling Libvirt "hey, this efi is capable of secure boot", nothing else.

jloeser replied:
Ok, then libvirt's documentation is a bit confusing:

According to https://libvirt.org/formatdomain.html#bios-bootloader (as already stated in https://github.com/fog/fog-libvirt/issues/128) the following enables UEFI+SB:
```
<!-- QEMU with automatic UEFI firmware and secure boot -->
...
<os firmware='efi'>
  <type>hvm</type>
  <loader secure='yes'/>
  <boot dev='hd'/>
</os>
...
```
But it *doesn't* mean, that `<loader secure='no'/>` would lead to UEFI without SB, correct?

And honestly I also do not fully understand why libvirt should be aware of the SB state. At the end it depends on the user to provide fitting boot files for successful boot up when SB is enabled. Maybe for other software running on top like `virt-manager`. Anyway ...

We should then follow https://libvirt.org/kbase/secureboot.html and provide (as written above) to explicitly enable or disable SB depending on the user's decision in the UI (and provide corresponding `secure`  value in addition for whatever it is good for).

stejskalleos replied:
> But it doesn't mean, that <loader secure='no'/> would lead to UEFI without SB, correct?

Exactly, that's why I started the discussion about the UI label. Secure Boot field in Foreman should be for enabling/disabling secure boot, and not only setting the `<loader secure='no'/>`.

stejskalleos commented:
Rebased, added @ekohl as author & tested with Fedora 39 UEFI + SecureBoot. Ready for review

stejskalleos commented:
@ekohl @jloeser Rebased and updated (Together with [FOG PR](https://github.com/fog/fog-libvirt/pull/141))
Now using
```xml
  <firmware>
    <feature enabled='yes' name='secure-boot'/>
    <feature enabled='yes' name='enrolled-keys'/>
  </firmware>
```

Tested with shimx64.efi & Fedora 39.

davispuh commented:
Nice work, I'm also interested in this.

nofaralfasi commented on app/views/compute_resources_vms/form/libvirt/_base.html.erb:
```diff
@@ -24,3 +24,27 @@
 
 <%= compute_specific_js(compute_resource, "nic_info") %>
+
+<%= select_f f, :os_firmware,
+                Foreman::Model::Libvirt.firmware_types,
+                :first,
+                :last,
+                {},
+                { :label => _("Firmware"),
+                  :label_size => "col-md-2",
+                  :onchange => "tfm.computeResource.libvirt.firmwareSelected(this);",
+                }
+%>
+<%
+  feature_attrs = ActiveSupport::HashWithIndifferentAccess.new(f.object.os_firmware_features)
+  is_bios = f.object.os_firmware == 'bios'
+%>
+
+<div id="os_firmware_features">
+<%= f.fields_for :os_firmware_features do |ff|
+    secure_field = checkbox_f(ff, 'secure-boot', { checked: feature_attrs['secure-boot'] == 'yes', label: _("Secure Boot"), disabled: is_bios }, 'yes', 'no')
+    keys_field = checkbox_f(ff, 'enrolled-keys', { checked: feature_attrs['enrolled-keys'] == 'yes', label: _("Enrolled keys"), disabled: is_bios }, 'yes', 'no')
+    secure_field + keys_field
+  end
+%>
+</div>
```
I think we should hide these options when editing a host and display them only when creating a new host. Since there is no way to edit these fields and we don’t know the actual values (we’re displaying the default ones), I don’t see a reason to include them in the edit form.

ekohl commented on app/models/compute_resources/foreman/model/libvirt.rb:
```diff
@@ -291,7 +313,12 @@ def vm_instance_defaults
                          :password => random_password(console_password_length(display_type)),
-                         :port     => '-1' }
+                         :port     => '-1' },
+        :os_firmware => 'efi',
```
This uses `os_firmware` and VMware uses `firmware`. That's inconsistent.

ekohl reviewed (changes requested):
In https://github.com/theforeman/foreman/pull/9965 I copied the VMware pattern to automatically detect the firmware based on the PXE loader. That should determine if secure boot it used by default.

nofaralfasi committed:
Refactor Secure Boot support for Libvirt

- Updated the Libvirt VM creation form.
- Added a new firmware type for Secure Boot.
- Enable `enrolled-keys` by default when Secure Boot is activated.
- Removed unnecessary methods from the Libvirt model.
- Moved firmware-related methods to the ComputeResource model
  for shared use between VMware and Libvirt.

nofaralfasi commented:
Requires: https://github.com/fog/fog-libvirt/pull/155.

nofaralfasi commented:
I have another idea on how to implement this. Moving it to WIP until it's ready.

nofaralfasi commented:
While determining the firmware type in Libvirt is straightforward when creating a VM, it can be more complex for existing VMs. The firmware can be specified in various places, such as `os.firmware`, `os.loader.type`, or it might not be specified at all. Additionally, there are factors to consider, such as the machine returning different values depending on whether it is running or not. For example, the loader attribute may only be available when the machine is running. Similarly, the Secure Boot feature can be enabled using `os.firmware.features` with `secure-boot` set to `yes`, but there are other methods to enable Secure Boot as well (refer to [Libvirt Secure Boot](https://libvirt.org/kbase/secureboot.html) and [Libvirt BIOS/Bootloader Documentation](https://libvirt.org/formatdomain.html#bios-bootloader) for more details).

Given these complexities, we have two options: either refrain from displaying the firmware type for existing machines, or display it with the understanding that the information may not always be accurate.

nofaralfasi committed:
Refactor Libvirt Firmware Handling for Better Testability

- Moved firmware logic into dedicated methods.
- Updated tests for improved coverage and workflow alignment.

nofaralfasi commented on app/models/compute_resources/foreman/model/libvirt.rb:
```diff
@@ -326,5 +330,26 @@ def validate_volume_capacity(vol)
       end
     end
+
+    def handle_automatic_firmware(opts)
+      # This value comes from PxeLoaderSupport#firmware_type
+      firmware_type = opts.delete(:firmware_type).to_s
+
+      # Handle automatic firmware setting based on PXE Loader
+      # if no PXE Loader is set, we will set it to bios by default
+      if opts[:firmware] == 'automatic'
+        opts[:firmware] = firmware_type.presence || 'bios'
+      end
+    end
```
This code is also used in https://github.com/theforeman/foreman/pull/10225/files#diff-be4ac4f95deb530eb0fbcc834a4940573ba45b8d5923433305ac3aa1e0703328R501-R508. 
I suggest moving this method into `app/models/compute_resource.rb` for shared use between VMware and Libvirt.

nofaralfasi commented:
Closing this PR in favor of a new one: #10321. The new PR addresses the same issue with updated changes. Please continue the discussion and review there.
